packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java

```diff
@@ -0,0 +1,248 @@
+
+      Uri imageUri = FileProvider.getUriForFile(registrar, providerName, imageFile);
+      intent.putExtra(MediaStore.EXTRA_OUTPUT, imageUri);
+      grantUriPermissions(intent, imageUri);
+
+      registrar.startActivityForResult(intent, REQUEST_CODE_TAKE_WITH_CAMERA);
+    }
+  }
+
+  private File createTemporaryWritableImageFile() {
+    String filename = UUID.randomUUID().toString();
+    File storageDirectory = registrar.getExternalFilesDir(Environment.DIRECTORY_PICTURES);
+    File image;
+
+    try {
+      image = File.createTempFile(filename, ".png", storageDirectory);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    return image;
+  }
+
+  private void grantUriPermissions(Intent intent, Uri imageUri) {
+    PackageManager packageManager = registrar.getPackageManager();
+    List<ResolveInfo> compatibleActivities =
+        packageManager.queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);
+
+    for (ResolveInfo info : compatibleActivities) {
+      registrar.grantUriPermission(
+          info.activityInfo.packageName,
+          imageUri,
+          Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
+    }
+  }
+
+  @Override
+  public boolean onRequestPermissionsResult(
+      int requestCode, String[] permissions, int[] grantResults) {
+    if (requestCode == REQUEST_IMAGE_PICKER_PERMISSIONS) {
+      boolean bothPermissionsGranted =
+          grantResults[0] == PackageManager.PERMISSION_GRANTED
+              && grantResults[1] == PackageManager.PERMISSION_GRANTED;
+
+      if (bothPermissionsGranted) {
+        executePendingOperation();
+      } else {
+        pendingResult.error(
+            "no_permissions", "image_picker plugin requires camera permissions", null);
+        clearMethodCallAndResult();
+      }
+      return true;
+    }
+    return false;
+  }
+
+  private void executePendingOperation() {
+    if (pendingPickImageOperation != null) {
+      switch (pendingPickImageOperation) {
+        case CHOOSE_FROM_GALLERY:
+          launchPickImageFromGalleryIntent();
+          break;
+        case TAKE_WITH_CAMERA:
+          launchTakeImageWithCameraIntent();
+          break;
+      }
+
+      pendingPickImageOperation = null;
+    }
+  }
+
+  @Override
+  public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
+    if (requestCode == REQUEST_CODE_CHOOSE_FROM_GALLERY) {
+      return handleChoosePictureResult(resultCode, data);
+    } else if (requestCode == REQUEST_CODE_TAKE_WITH_CAMERA) {
+      return handleTakePictureResult(resultCode);
+    }
+    return false;
+  }
+
+  private boolean handleChoosePictureResult(int resultCode, Intent data) {
+    if (resultCode == Activity.RESULT_OK && data != null) {
+      String path = FileUtils.getPathFromUri(registrar, data.getData());
+      handleResult(path);
+      return true;
+    } else if (resultCode != Activity.RESULT_CANCELED) {
```

Comment:
So on Activity.RESULT_CANCELED we don't send anything back to the dart side? Will that not cause the app to hang because somebody is awaiting a result there?